File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ExpressionVisitor.java
/*
*  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
*
*  WSO2 Inc. licenses this file to you under the Apache License,
*  Version 2.0 (the "License"); you may not use this file except
*  in compliance with the License.
*  You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing,
*  software distributed under the License is distributed on an
*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
*  KIND, either express or implied.  See the License for the
*  specific language governing permissions and limitations
*  under the License.
*/
package org.ballerinalang.model.expressions;

import org.ballerinalang.model.expressions.variablerefs.FieldBasedVarRefExpr;
import org.ballerinalang.model.expressions.variablerefs.IndexBasedVarRefExpr;
import org.ballerinalang.model.expressions.variablerefs.SimpleVarRefExpr;
import org.ballerinalang.model.expressions.variablerefs.XMLAttributesRefExpr;

/**
 * @since 0.92
 */
public interface ExpressionVisitor {

    Expression visit(BasicLiteral basicLiteral);

    Expression visit(NullLiteral nullLiteral);

    Expression visit(SimpleVarRefExpr simpleVarRefExpr);

    Expression visit(FieldBasedVarRefExpr fieldBasedVarRefExpr);

    Expression visit(IndexBasedVarRefExpr indexBasedVarRefExpr);

    Expression visit(UnaryExpression unaryExpression);

    Expression visit(AddExpression addExpr);

    Expression visit(SubtractExpression subtractExpression);

    Expression visit(MultExpression multExpression);

    Expression visit(DivideExpr divideExpr);

    Expression visit(ModExpression modExpression);

    Expression visit(AndExpression andExpression);

    Expression visit(OrExpression orExpression);

    Expression visit(EqualExpression equalExpression);

    Expression visit(NotEqualExpression notEqualExpression);

    Expression visit(GreaterEqualExpression greaterEqualExpression);

    Expression visit(GreaterThanExpression greaterThanExpression);

    Expression visit(LessEqualExpression lessEqualExpression);

    Expression visit(LessThanExpression lessThanExpression);

    Expression visit(FunctionInvocationExpr functionInvocationExpr);

    Expression visit(ActionInvocationExpr actionInvocationExpr);

    Expression visit(TypeCastExpression typeCastExpression);

    Expression visit(TypeConversionExpr typeConversionExpression);

    Expression visit(ArrayInitExpr arrayInitExpr);

    Expression visit(RefTypeInitExpr refTypeInitExpr);

    Expression visit(ConnectorInitExpr connectorInitExpr);

    Expression visit(StructInitExpr structInitExpr);

    Expression visit(MapInitExpr mapInitExpr);

    Expression visit(JSONInitExpr jsonInitExpr);

    Expression visit(JSONArrayInitExpr jsonArrayInitExpr);

    Expression visit(XMLLiteral xmlLiteral);

    Expression visit(XMLElementLiteral xmlElement);

    Expression visit(XMLCommentLiteral xmlComment);

    Expression visit(XMLTextLiteral xmlText);

    Expression visit(XMLPILiteral xmlPI);

    Expression visit(XMLAttributesRefExpr xmlAttributesRefExpr);

    Expression visit(XMLQNameExpr xmlQNameRefExpr);

    Expression visit(XMLSequenceLiteral xmlSequence);

    Expression visit(LambdaExpression lambdaExpr);
}

File: composer/packages/diagram/src-old/plugins/ballerina/visitors/expression-visitor.js
/**
 * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import ASTVisitor from './ast-visitor';
import ASTFactory from '../ast/ast-factory';

class ExpressionVisitor extends ASTVisitor {
    constructor(args) {
        super(args);
    }

    canVisitLeftOperandExpression(expression) {
        return false;
    }

    beginVisitLeftOperandExpression(expression) {
    }

    visitLeftOperandExpression(expression) {
    }

    endVisitLeftOperandExpression(expression) {
    }

    canVisitRightOperandExpression(expression) {
        return false;
    }

    beginVisitRightOperandExpression(expression) {
    }

    visitRightOperandExpression(expression) {
    }

    endVisitRightOperandExpression(expression) {
    }

    canVisitSimpleVariableReferenceExpression(expression) {
        return false;
    }

    beginVisitSimpleVariableReferenceExpression(expression) {
    }

    visitSimpleVariableReferenceExpression(expression) {
    }

    endVisitSimpleVariableReferenceExpression(expression) {
    }

    canVisitTypeCastExpression(expression) {
        return false;
    }

    beginVisitTypeCastExpression(expression) {
    }

    visitTypeCastExpression(expression) {
    }

    endVisitTypeCastExpression(expression) {
    }

    canVisitTypeConversionExpression(expression) {
        return false;
    }

    beginVisitTypeConversionExpression(expression) {
    }

    visitTypeConversionExpression(expression) {
    }

    endVisitTypeConversionExpression(expression) {
    }

    /**
     * @param node {ASTNode}
     */
    visitExpression(node) {
        if (ASTFactory.isSimpleVariableReferenceExpression(node)) {
            return this.visitSimpleVariableReferenceExpression(node);
        } else if (ASTFactory.isTypeCastExpression(node)) {
            return this.visitTypeCastExpression(node);
        } else if (ASTFactory.isTypeConversionExpression(node)) {
            return this.visitTypeConversionExpression(node);
        } else if (ASTFactory.isLeftOperandExpression(node)) {
            return this.visitLeftOperandExpression(node);
        } else if (ASTFactory.isRightOperandExpression(node)) {
            return this.visitRightOperandExpression(node);
        }
    }

    /**
     * @param node {ASTNode}
     */
    canVisitExpression(node) {
        if (ASTFactory.isSimpleVariableReferenceExpression(node)) {
            return this.canVisitSimpleVariableReferenceExpression(node);
        } else if (ASTFactory.isTypeCastExpression(node)) {
            return this.canVisitTypeCastExpression(node);
        } else if (ASTFactory.isTypeConversionExpression(node)) {
            return this.canVisitTypeConversionExpression(node);
        } else if (ASTFactory.isLeftOperandExpression(node)) {
            return this.canVisitLeftOperandExpression(node);
        } else if (ASTFactory.isRightOperandExpression(node)) {
            return this.canVisitRightOperandExpression(node);
        }
    }

    /**
     * @param node {ASTNode}
     */
    beginVisitExpression(node) {
        if (ASTFactory.isSimpleVariableReferenceExpression(node)) {
            this.beginVisitSimpleVariableReferenceExpression(node);
        } else if (ASTFactory.isTypeCastExpression(node)) {
            this.beginVisitTypeCastExpression(node);
        } else if (ASTFactory.isTypeConversionExpression(node)) {
            this.beginVisitTypeConversionExpression(node);
        } else if (ASTFactory.isLeftOperandExpression(node)) {
            return this.beginVisitLeftOperandExpression(node);
        } else if (ASTFactory.isRightOperandExpression(node)) {
            return this.beginVisitRightOperandExpression(node);
        }
    }

    /**
     * @param node {ASTNode}
     */
    endVisitExpression(node) {
        if (ASTFactory.isSimpleVariableReferenceExpression(node)) {
            return this.endVisitSimpleVariableReferenceExpression(node);
        } else if (ASTFactory.isTypeCastExpression(node)) {
            return this.endVisitTypeCastExpression(node);
        } else if (ASTFactory.isTypeConversionExpression(node)) {
            return this.endVisitTypeConversionExpression(node);
        } else if (ASTFactory.isLeftOperandExpression(node)) {
            return this.endVisitLeftOperandExpression(node);
        } else if (ASTFactory.isRightOperandExpression(node)) {
            return this.endVisitRightOperandExpression(node);
        }
    }
}

export default ExpressionVisitor;
